File: clients/android/app/src/main/java/com/clipbridge/ReconnectPolicy.kt
package com.clipbridge

/** What the foreground reconnect watchdog should do on one tick. */
internal enum class ForegroundReconnectAction {
    /** Leave the transport alone: connected, backgrounded, or standby owns it. */
    NONE,

    /** There is no client at all — build one. */
    START_CLIENT,

    /** A client exists but isn't connected — ask it to redial immediately. */
    RECONNECT_NOW,
}

/**
 * Android hands the app no dependable background lifetime, so a session that
 * died while the user was away is routine. Whenever the app is in front and
 * the link is not up, we act rather than report.
 *
 * Two cases deliberately do nothing:
 *
 *   - Screen-off standby (`reconnectIdleMode`) owns the transport; reconnecting
 *     underneath it would undo the hard suspend. Only reachable as a brief race
 *     between a lock and the watchdog tick.
 *   - Backgrounded, which the watchdog also checks before sleeping — this is
 *     the belt to that suspenders, since the flag can flip mid-tick.
 */
internal fun foregroundReconnectAction(
    hostAppForeground: Boolean,
    reconnectIdleMode: Boolean,
    hasClient: Boolean,
    isConnected: Boolean,
): ForegroundReconnectAction = when {
    !hostAppForeground -> ForegroundReconnectAction.NONE
    isConnected -> ForegroundReconnectAction.NONE
    reconnectIdleMode -> ForegroundReconnectAction.NONE
    !hasClient -> ForegroundReconnectAction.START_CLIENT
    else -> ForegroundReconnectAction.RECONNECT_NOW
}

/**
 * Gap before the watchdog's next tick. Ticks that actually retried double the
 * gap up to [maxMs] so a genuinely dead network doesn't turn a long foreground
 * session into a dial loop; anything else (connected, or standby holding the
 * transport) drops back to [minMs] so the next real outage is caught quickly.
 */
internal fun nextForegroundReconnectDelayMs(
    currentMs: Long,
    retried: Boolean,
    minMs: Long,
    maxMs: Long,
): Long = if (retried) (currentMs * 2).coerceAtMost(maxMs) else minMs

File: clients/macos/Sources/ClipBridgeApp/ReconnectPolicy.swift
import Foundation

/// What the connection watchdog should do on one tick.
enum ReconnectAction: Equatable {
    /// The link is up — leave the core alone.
    case none

    /// There is no core client at all. The constructor threw (or the key
    /// failed to decode) and nothing ever retried it, so the menu sat on
    /// whatever status was showing with nothing behind it.
    case startClient

    /// A client exists and is still reporting state, it just isn't connected:
    /// cut its backoff short instead of waiting out up to 30s of it.
    case reconnectNow

    /// A client exists but the core has gone quiet while disconnected — its
    /// worker thread is dead or wedged. Nothing short of a new client helps.
    case restartClient
}

/// Tuning for the watchdog. Mirrors the Android foreground watchdog
/// (5s→30s) so both hosts recover on the same cadence.
enum ReconnectTuning {
    static let minDelay: TimeInterval = 5
    static let maxDelay: TimeInterval = 30

    /// How long the core may report nothing at all, while not connected,
    /// before we treat its worker as gone. A live worker emits a state on
    /// every dial attempt, and its slowest possible cycle is a 15s connect
    /// cap plus a 30s backoff — so silence past this is not a slow network,
    /// it is nobody home.
    static let coreSilenceLimit: TimeInterval = 120
}

/// Decide one watchdog tick. Pure so the escalation order (retry → replace)
/// is testable without a menu bar, a network, or a core.
///
/// `workerAlive` is the core's own answer (`Client.isRunning()`); a worker
/// that unwound reports its crash and then never speaks again, so redialing
/// it is pointless — only a new client recovers. Silence is the same verdict
/// arrived at the slow way, for a worker that is stuck rather than gone.
func reconnectAction(
    hasClient: Bool,
    workerAlive: Bool,
    isConnected: Bool,
    sinceLastState: TimeInterval,
    silenceLimit: TimeInterval = ReconnectTuning.coreSilenceLimit
) -> ReconnectAction {
    if isConnected { return .none }
    guard hasClient else { return .startClient }
    if !workerAlive { return .restartClient }
    return sinceLastState >= silenceLimit ? .restartClient : .reconnectNow
}

/// Gap before the next tick. Ticks that actually retried double the gap up to
/// `max` so a genuinely dead network doesn't turn into a dial loop; a tick
/// with nothing to do drops back to `min` so the next real outage is caught
/// quickly.
func nextWatchdogDelay(
    current: TimeInterval,
    retried: Bool,
    min minDelay: TimeInterval = ReconnectTuning.minDelay,
    max maxDelay: TimeInterval = ReconnectTuning.maxDelay
) -> TimeInterval {
    retried ? Swift.min(current * 2, maxDelay) : minDelay
}
